tests/beta/interfaces/test_tape_tf.py

```diff
@@ -0,0 +1,741 @@
+
+        dev = qml.device("default.qubit.tf", wires=2)
+        x = tf.Variable(0.543, dtype=tf.float64)
+        y = tf.Variable(-0.654, dtype=tf.float64)
+
+        with tf.GradientTape() as tape:
+            with QuantumTape() as qtape:
+                qml.RX(x, wires=[0])
+                qml.RY(y, wires=[1])
+                qml.CNOT(wires=[0, 1])
+                probs(wires=[0])
+                probs(wires=[1])
+
+            res = qtape.execute(dev)
+
+        expected = np.array(
+            [
+                [tf.cos(x / 2) ** 2, tf.sin(x / 2) ** 2],
+                [(1 + tf.cos(x) * tf.cos(y)) / 2, (1 - tf.cos(x) * tf.cos(y)) / 2],
+            ]
+        )
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+        res = tape.jacobian(res, [x, y])
+        expected = np.array(
+            [
+                [
+                    [-tf.sin(x) / 2, tf.sin(x) / 2],
+                    [-tf.sin(x) * tf.cos(y) / 2, tf.cos(y) * tf.sin(x) / 2],
+                ],
+                [
+                    [0, 0],
+                    [-tf.cos(x) * tf.sin(y) / 2, tf.cos(x) * tf.sin(y) / 2],
+                ],
+            ]
+        )
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+    def test_ragged_differentiation(self, monkeypatch, tol):
+        """Tests correct output shape and evaluation for a tape
+        with prob and expval outputs"""
+        dev = qml.device("default.qubit.tf", wires=2)
+        x = tf.Variable(0.543, dtype=tf.float64)
+        y = tf.Variable(-0.654, dtype=tf.float64)
+
+        def _asarray(args, dtype=tf.float64):
+            res = [tf.reshape(i, [-1]) for i in args]
+            res = tf.concat(res, axis=0)
+            return tf.cast(res, dtype=dtype)
+
+        # we need to patch the asarray method on the device
+        monkeypatch.setattr(dev, "_asarray", _asarray)
+
+        with tf.GradientTape() as tape:
+            with QuantumTape() as qtape:
+                qml.RX(x, wires=[0])
+                qml.RY(y, wires=[1])
+                qml.CNOT(wires=[0, 1])
+                expval(qml.PauliZ(0))
+                probs(wires=[1])
+
+            res = qtape.execute(dev)
+
+        expected = np.array(
+            [tf.cos(x), (1 + tf.cos(x) * tf.cos(y)) / 2, (1 - tf.cos(x) * tf.cos(y)) / 2]
+        )
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+        res = tape.jacobian(res, [x, y])
+        expected = np.array(
+            [
+                [-tf.sin(x), -tf.sin(x) * tf.cos(y) / 2, tf.cos(y) * tf.sin(x) / 2],
+                [0, -tf.cos(x) * tf.sin(y) / 2, tf.cos(x) * tf.sin(y) / 2],
+            ]
+        )
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+    def test_sampling(self):
+        """Test sampling works as expected"""
+        dev = qml.device("default.qubit.tf", wires=2, shots=10)
+
+        with tf.GradientTape() as tape:
+            with QuantumTape() as qtape:
+                qml.Hadamard(wires=[0])

```

Comment:
How about making the test deterministic (e.g. applying `qml.PauliX(0)` and having `sample(qml.PauliZ(0))` or comparable) and also checking the samples?

Reply:
I'm not sure it's necessary to check the sample values here; I care only that the machinary works, and that the interface properly returns (a) the right type, and (b) the right shape.

I think checking the samples is more the job of the device tests.

So in a sense, while there is a stochastic computation happening behind the scenes, this test kind of _is_ already deterministic, as we are only checking deterministic quantities!